### PR TITLE
Oracle support for storage and system store

### PR DESCRIPTION
**Pull request by eisendle**
This pull request are the first steps for supporting Oracle as storage provider and system store.
Im still in discussion with @veikkoeeva about some remaining issues, but this pull request adds parts where additional discussion is not needed.

### CODE REVIEW

**Comment by sergeybykov** [comment]
@veikkoeeva Do you approve this change?

**Comment by veikkoeeva** [comment]
@eisendle I'll take a better look at tomorrow, but a few notes with a glance of it:
- Please, add comments as in other scripts and try to keep the tables and queries in the same order. This makes it easier to keep the scripts synchronized. 
- About the auto-incrementing value, it would be much preferrable to keep a similar increment-by-one structure as the other scripts due to consistency (added note: https://stackoverflow.com/questions/10613846/create-table-with-sequence-nextval-in-oracle).

**Comment by eisendle** [comment]
@veikkoeeva 

- I will change the order of the statements in order that they are in sync with  the other scripts and add comments, when the other open issues are resolved.
- The script is already using a sequence which creates auto-increment IDs. I did not use the IDENTITY feature by purpose because this would push the minimum required oracle version to 12 as the IDENTITY support was added in version 12. See the SO article you mentioned.

**Comment by veikkoeeva** [comment]
@eisendle 

1. On the order of statements, sounds very good.
2. A very good point. A related question, would it be difficult to make it an increment by one as the other scripts are doing? I think it's OK to use trigger too, but going to the direction of being more uniform with the others.

Maybe useful background information on Storage related parts from the implementation notes: https://github.com/dotnet/orleans/pull/1682#r71064076. Since the potentially large strings are hashed, the result resembles that of reverse index in Oracle. The idea here is that the index doesn't need to be ordered and should be fairly maintenace free and if possible, compression could be used.

A question: Does Oracle have a free to use, lightweight version of the DB engine to use in testing? The other DBs are basically added to the matrix with default settings so if one has them installed, the tests show wither red or green instead of yellow (inconclusive).

**Comment by eisendle** [comment]
@veikkoeeva 

ad 2.
The sequence and the trigger already mock an auto increment column, so the  Id is incremented by one on each insert.

Regarding the index: I will look into that, as I am no oracle guru I have to dig into that topic

There is indeed an lightweight version for oracle:  [Oracle Express](http://www.oracle.com/technetwork/database/database-technologies/express-edition/downloads/index.html) I am using it myself for testing.

Another question from my side how is the test setup for the unit tests? Is there an installation of MSSQL,Postgres and Mysql on the build machines? Or are the database related tests skipped on the build machines?

A quick update on the state of this PR: I making good progress, Storage, Reminders and Membership are working.

Open work:

- Change the order of the statements in the create script in order to be sync with other scripts
- Add comments to the SQL script which clarifies why some things are implemented in a certain way
- Possibly optimize the index on the storage tables
- Test multi insert of statistics
- Test whether metrics persistence is working as expected

**Comment by jdom** [comment]
> Another question from my side how is the test setup for the unit tests? Is there an installation of MSSQL,Postgres and Mysql on the build machines? Or are the database related tests skipped on the build machines?

Unfortunately no. Only SQL Server is installed in our build machines (not in Jenkins, but in our VSO, that we run after each commit is merged to master, if there's risk, sometimes we proactively run a PR in that environment too).
For the rest we don't run any automated tests ourselves, and it's up to the contributors, we just facilitate the space for these 3rd party integrations to be developed and released. If someone could contribute some automation that relies on Docker or something to spin up a test database, it would be awesome.

**Comment by galvesribeiro** [comment]
Was thinking about that few days ago... @jdom do you know if the Jenkins build agent has docker installed? I know VSTS has it, so spin an instance of a DB (or whatever other dependency) as a container is easy there. 

CC @mmitche as he may know the answer.

**Comment by jdom** [comment]
Yes, docker is installed

**Comment by galvesribeiro** [comment]
Ok, so we are able to run it. Lets talk offline about it when you have time. We should be able to easily automate the invocation of those containers from the test framework and perform the tests that requires external dependencies like this one. Ping me when you have some time.

**Comment by veikkoeeva** [comment]
@galvesribeiro If there were a nice way to setup a Docker environment from .NET Core 2.0 in code, that would be great as it'd make setting up a testing environment more self-contained. :) Though I suppose this is more like a footnote at this point.

**Comment by veikkoeeva** [comment]
@eisendle If you need help setting up tests codewise if you have access to Oracle DB, please do ask here or in Gitter. This is good piece of work you're doing and we'd like you not to sweat too much on it.

**Comment by galvesribeiro** [comment]
@veikkoeeva yes, there is. I've been playing with Docker+C# integration for a while. We can definitively put up and tear down Docker containers inside the TestFixture ctor and Dispose() respectively. Will align with @jdom once he find time. Shouldn't take long as we have few of those dependencies.

File: src/OrleansSQLUtils/Storage/DbStoredQueries.cs [hunk 1]
```diff
@@ -319,19 +321,22 @@ internal List<ICounter> Counters
...
-                    Add(nameof(value.IsOverloaded), value.IsOverloaded);
-                    Add(nameof(value.ClientCount), value.ClientCount);
-                }
+                AddCoreMetricsParams(siloPerformanceMetrics);
+                Add(nameof(siloPerformanceMetrics.ActivationCount), siloPerformanceMetrics.ActivationCount);
+                Add(nameof(siloPerformanceMetrics.RecentlyUsedActivationCount), siloPerformanceMetrics.RecentlyUsedActivationCount);
+                Add(nameof(siloPerformanceMetrics.RequestQueueLength), siloPerformanceMetrics.RequestQueueLength);
+
+                //Yet another fix for Oracle Data Provider: ODP.net doesn´t support boolean parameters
+                if(mapIsOverloadedAsBoolean)
```
[the remark this answers is not in the document]
Comment (eisendle) [reply src/OrleansSQLUtils/Storage/DbStoredQueries.cs]: done

**Comment by eisendle** [comment]
@veikkoeeva 
The oracle integration looked good until today when I discovered a problem. When writing large values (>32kb) into the JsonPayload column of the storage table, oracle is complaining that the VARCHAR value is too large. In the storage table JsonPayload is mapped as CLOB which supports up 4GB of content. The reason for the problem is that the DBParameter for JSONPayload is mapped as varchar. I have tried:

- Convert the JsonPayload query parameter to a CLOB with the function TO_CLOB in the query -> NOT WORKING
- Try different System.Data.DbTypes for the parameter in order to outwit oracle -> NOT WORKING
- Setting the OracleDbType of the OracleParameter explicitly to OracleDbType.Clob -> WORKS

However this solution is unfavorable as it would need applying reflection which breaks the ado.net abstraction. I´m afraid this is the only solution and would like discuss whether we want such a hack in the codebase. I would implement some kind of Interceptor which is invoked  before a command is executed in the RelationStorage. Other providers wouldn´t be affected by this change as a "NullInterceptor" would be created for the other providers.
When this approach is rejected, I will close this pull request because this issue is a show stopper. Also I have to admit I´m tired of working around oracle "specifics" and oddities.

**Comment by veikkoeeva** [comment]
@eisendle I think it's OK to state this currently as a limitation and document it along with potential solutions and leave it as is for now. This looks like an edge case currently. I tend to think Oracle is problematic once one goes outside of the "basic zone" and this looks like such.

**Comment by veikkoeeva** [comment]
@eisendle Also, don't burn yourself down. Any progress is better than no progress. It's great you're applying some real stress testing to the solution. 👍

**Comment by eisendle** [comment]
@veikkoeeva I didn´t want to leave the oracle orleans support with this limitations, so tried to bypass the limitations with the latest commit. Like I mentioned previously some kind of Interceptor class could set the required Oracle related properties via reflection. 
I didn´t like the usage of reflection so the added OracleDatabaseCommandInterceptor class compiles lambda expressions on runtime, so that there is only a one time reflection penalty.
Also the Interceptor encapsulates all oracle specifics in one class.
The interceptor is invoked for every IDbCommand instance before the query is executed. 
All other data providers are unaffected as there is another implemenation of IDatabaseCommandInterceptor which does nothing.
Also there are no oracle specific fixes in DbStoredQueries anymore, all moved into the interceptor. 
Can you pls take a look on the changes?

**Comment by eisendle** [comment]
Open work:

- [ ] Remove Obsolete supportsBooleanParameter field from DBConstantsStore
- [ ] Rename oracle functions -> Remove Key suffix from function names

**Comment by veikkoeeva** [comment]
@eisendle You are labouring! 👍 Looks good without anything serious. Maybe little less redundancy in naming (e.g. is the phrase `database` the correct one or even needed one since it's indeed ADO.NET storage already). One could think of the granulatiry of interceptors (i.e. per call, per connection string, per provider), but let's go what's simplest as of now (I might want to add `IAsyncEnumerable` and compiled delegates in the future).

I'm in www.slush.org and related things this week, so rather unavailable, but you can ping me or @shayhatsor or anyone else naturally. And after that I'm online better.

**Comment by eisendle** [comment]
@veikkoeeva Can you pls take a look at the PR. Looks good IMO and ready to merge.

**Comment by veikkoeeva** [comment]
@eisendle In about 10-12 hours I will. 👍

**Comment by veikkoeeva** [comment]
@eisendle I find the code is OK, no reason to hold because of that. I have trouble running the tests (it says could not copy `xunit.abstractions.dll`) and I'm short on time finding out why and fixing it. If you have run the tests for Oracle and checked at least SQL Server works too still, I'm inclined to see so does MySQL and PostgreSQL too.

Would this be OK, @sergeybykov?

**Comment by veikkoeeva** [comment]
Ah, but I see, for instance https://ci.dot.net/job/dotnet_orleans/job/master/job/functional_prtest/693/testReport/UnitTests.StorageTests.Relational/SqlServerStorageTests/ still pass, so life is good. :)

**Comment by jdom** [comment]
The reason for that error is that you need to configure the test runner in VS like so (after opening the solution):
`Test` -> `Test Settings` -> Uncheck `Keep Test Execution Engine running`
`Test` -> `Test Settings` -> `Default Processor Architecture` -> Check `X64`

Then either restart VS, or go to the task manager and kill the processes that start with `vstest.`, and build

**Comment by sergeybykov** [comment]
Thank you so much, @eisendle, for putting all the effort into this! Oracle is obviously an important database to support.

@veikkoeeva, thank you for reviewing and guiding this effort!

**Comment by dVakulen** [comment]
Also thank you @jdom for test runner configuration advice.

**Comment by jdom** [comment]
Yes, thanks a lot to all, great effort! :)
BTW, put the tips for running the tests in VS into our readme, so this tribal knowledge is documented somewhere

**Pull request merged**